File: Python/059__Spiral_Matrix_II.py
class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        val = 0
        result = [[0] * n for _ in range(n)]
        direct = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        i, j = 0, -1
        cur_dir = 0
        while val < n * n:
            val += 1
            next_i, next_j = i + direct[cur_dir][0], j + direct[cur_dir][1]
            if 0 <= next_i < n and 0 <= next_j < n and result[next_i][next_j] == 0:
                result[next_i][next_j] = val
            else:
                cur_dir = (cur_dir + 1) % 4
                next_i, next_j = i + direct[cur_dir][0], j + direct[cur_dir][1]
                result[next_i][next_j] = val
            i, j = next_i, next_j
        return result

File: Java/059__Spiral_Matrix_II.java
class Solution {
    public int[][] generateMatrix(int n) {
        if (n == 0)
            return new int[][]{};
        int[][] result = new int[n][n];
        int upBounds = 0, downBounds = n - 1;
        int rightBounds = n - 1, leftBounds = 0;
        int count = 1;
        while (count <= n * n) {
            for (int i = leftBounds; i <= rightBounds; i++)
                result[upBounds][i] = count++;
            for (int i = upBounds + 1; i <= downBounds; i++)
                result[i][rightBounds] = count++;
            for (int i = rightBounds - 1; i >= leftBounds; i--)
                result[downBounds][i] = count++;
            for (int i = downBounds - 1; i > upBounds; i--)
                result[i][leftBounds] = count++;
            upBounds++;
            downBounds--;
            rightBounds--;
            leftBounds++;
        }

        return result;
    }
}
